packages/firestore/src/core/transaction.ts

```diff
@@ -183,6 +116,84 @@ export class Transaction {
     });
   }
 
+  private recordVersion(doc: MaybeDocument): void {
+    let docVersion: SnapshotVersion;
+
+    if (doc instanceof Document) {
+      docVersion = doc.version;
+    } else if (doc instanceof NoDocument) {
+      // For deleted docs, we must use baseVersion 0 when we overwrite them.
+      docVersion = SnapshotVersion.forDeletedDoc();
+    } else {
+      throw fail('Document in a transaction was a ' + doc.constructor.name);
+    }
+
+    const existingVersion = this.readVersions.get(doc.key);
+    if (existingVersion !== null) {
+      if (!docVersion.isEqual(existingVersion)) {
+        // This transaction will fail no matter what.
+        throw new FirestoreError(
+          Code.ABORTED,
+          'Document version changed between two reads.'
+        );
+      }
+    } else {
+      this.readVersions = this.readVersions.insert(doc.key, docVersion);
+    }
+  }
+
+  private write(mutations: Mutation[]): void {
```

Comment:
Optional: In your new ordering you might put `write` at the bottom to make this flow more (generally a transaction reads a value (which records a version), we generate the precondition for any write and then actually write.